### PR TITLE
Update selective disclosure functions and their use in ecdsa-sd sign, derive, verify.

### PR DESCRIPTION
**Pull request by dlongley**
This PR updates the underlying selective disclosure functions to remove the need to use the JSON-LD framing primitive and to remove the special treatment of `credentialSubject` for VCs. It accomplishes this by using JSON pointers for selection (filtering and grouping) and skolemization of expanded (and subsequently compacted) JSON-LD documents instead of direct skolemization of N-Quads. Grouping and filter functions have been removed and replaced with a single `canonicalizeAndGroup` function, and the previous `jsonPointersToFrame` function has been adjusted to `selectJsonLd` as it now performs selection directly on skolemized, compact JSON-LD documents instead of producing a frame as an intermediate step.

The formats / expression of proof data in ecdsa-sd is unchanged, only the internals used to implement it are updated.

A few other corrections were made around label factory function production.

Some editorial clean up still needs to be done around the use of `<var>`, `<em>` and quoting variables using backticks, but this is not a new issue.

***
<a href="https://pr-preview.s3.amazonaws.com/dlongley/vc-di-ecdsa/pull/27.html" title="Last updated on Aug 16, 2023, 2:58 PM UTC (debfb6a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-di-ecdsa/27/6f9b3fa...dlongley:debfb6a.html" title="Last updated on Aug 16, 2023, 2:58 PM UTC (debfb6a)">Diff</a>

### CODE REVIEW

**Review by Wind4Greg — APPROVED** [review]
This is a nice improvement to the internal selective disclosure processing (both mandatory reveal, and selective reveal) to remove the dependency on JSON-LD framing. Key functionality is encapsulated in the `canonicalizeAndGroup` function.

It should be noted that besides the ECDSA algorithms from FIPS-186-5 (2023) that the EdDSA algorithms (Ed25519, Ed448) can be used. In addition the SD primitives particularly `canonicalizeAndGroup` maybe applicable for applying BBS to VCs.

File: index.html [hunk 1]
```diff
@@ -1325,332 +1484,375 @@ <h4>jsonPointerToPaths</h4>
...
+options (such as a document loader) can be passed. The <var>document</var> is
+assumed to use a JSON-LD context that aliases "@id" and "@type" to "id" and
+"type", respectively, and to use only one <em>@context</em> property at the top
+level of the document. An object containing the created groups
+(<var>groups</var>), the skolemized, compact JSON-LD document
+(<var>skolemizedCompactDocument</var>) the skolemized, expanded JSON-LD
+document (<var>skolemizedExpandedDocument</var>), the deskolemized N-Quad
+strings (<var>deskolemizedNQuads</var>), the blank node label map
+(<var>labelMap</var>), and the canonical N-Quad strings <var>nquads</var> is
+produced as output.
```
Comment (TallTed) [inline index.html:1773]:
```suggestion
The required inputs are a compact JSON-LD document (<var>document</var>),
a label map factory function (<var>labelMapFactoryFunction</var>), and a map
of named group definitions (<var>groupDefinitions</var>). Additional custom
options (such as a document loader) can be passed. The <var>document</var> is
assumed to use a JSON-LD context that aliases `@id` and `@type` to `id` and
`type`, respectively, and to use only one <em>`@context`</em> property at the top
level of the document. An object containing the created groups
(<var>groups</var>), the skolemized compact JSON-LD document
(<var>skolemizedCompactDocument</var>), the skolemized expanded JSON-LD
document (<var>skolemizedExpandedDocument</var>), the deskolemized N-Quad
strings (<var>deskolemizedNQuads</var>), the blank node label map
(<var>labelMap</var>), and the canonical N-Quad strings <var>nquads</var>, is
produced as output.
```

File: index.html [hunk 2]
```diff
@@ -1325,332 +1484,375 @@ <h4>jsonPointerToPaths</h4>
...
+          <ol class="algorithm">
             <li>
-Note: Next generate matching and non-matching maps composed of original indexes
-to original N-Quads. The `groupResult` is different; it contains matching and
-non-matching maps using the `filteredNQuads` indexes. Both maps of indexes are
-useful to callers.
+Initialize <var>skolemizedExpandedDocument</var> and
+<var>skolemizedCompactDocument</var> to their associated values in the
+result of the algorithm in Section <a href="#skolemizecompactjsonld"></a>,
+passing <var>document</var> and any custom options.
```
Comment (TallTed) [inline index.html:1781]:
```suggestion
passing <var>document</var> with any custom options.
```

File: index.html [hunk 3]
```diff
@@ -1325,332 +1484,375 @@ <h4>jsonPointerToPaths</h4>
...
             </li>
             <li>
-Initialize `matching` to a new map.
+Initialize <var>deskolemizedNQuads</var> to the result of the
+algorithm in Section <a href="#todeskolemizednquads"></a>, passing
+<var>skolemizedExpandedDocument</var> and any custom options.
```
Comment (TallTed) [inline index.html:1786]:
```suggestion
<var>skolemizedExpandedDocument</var> with any custom options.
```

File: index.html [hunk 4]
```diff
@@ -1325,332 +1484,375 @@ <h4>jsonPointerToPaths</h4>
...
             </li>
             <li>
-Initialize `nonMatching` to a new map.
+Initialize <var>nquads</var> and <var>labelMap</var> to their associated
+values in the result of the algorithm in Section
+<a href="#labelreplacementcanonicalizenquads"></a>, passing
+<var>deskolemizedNQuads</var> as <var>nquads</var>,
+<var>labelMapFactoryFunction</var>, and any custom options.
```
Comment (TallTed) [inline index.html:1793]:
```suggestion
and <var>labelMapFactoryFunction</var>, with any custom options.
```

File: index.html [hunk 5]
```diff
@@ -1325,332 +1484,375 @@ <h4>jsonPointerToPaths</h4>
...
             </li>
             <li>
-Initialize filteredNonMatches to the values in `groupResult.nonMatching`.
+For each key (<var>name</var>) and value (<var>pointers</var>) entry in
+<var>groupDefinitions</var>:
+              <ol class="algorithm">
+                <li>
+Add an entry with a key of <var>name</var> and a value that is the result
+of the algorithm in Section <a href="#selectcanonicalnquads"></a>, passing
+<var>skolemizedCompactDocument</var> as <var>document</var>,
```
Comment (TallTed) [inline index.html:1805]:
```suggestion
<var>skolemizedCompactDocument</var> as <var>document</var>, with
```

File: index.html [hunk 6]
```diff
@@ -1325,332 +1484,375 @@ <h4>jsonPointerToPaths</h4>
...
+For each element (<var>nq</var>) and index (<var>index</var>) in
+<var>nquads</var>:
+                  <ol class="algorithm">
+                    <li>
+Create a map entry, <var>entry</var>, with a key of <var>index</var> and a
+value of <var>nq</var>.
+                    </li>
+                    <li>
+If <var>selectedNQuads</var> includes <var>nq</var> then add <var>entry</var>
+to <var>matching</var>, otherwise add <var>entry</var> to <var>nonMatching</var>.
```
Comment (TallTed) [inline index.html:1841]:
```suggestion
to <var>matching</var>; otherwise, add <var>entry</var> to <var>nonMatching</var>.
```

File: index.html [hunk 7]
```diff
@@ -1819,65 +2021,53 @@ <h4>createDisclosureData</h4>
 algorithm used in the signature algorithm, i.e., SHA-256 for a P-256 curve.
             </li>
             <li>
-Initialize `nquads` to the result of calling the algorithm in Section
-<a href="#hmacidcanonize"></a>, passing `document`, `hmac`, and any custom
-JSON-LD API options as parameters. Note: This step transforms the document into
-an array of canonical N-Quads with pseudorandom blank node identifiers based on
-`hmac`.
+Initialize `labelMapFactoryFunction` to the result of callign the
```
Comment (TallTed) [inline index.html:2024]:
```suggestion
Initialize `labelMapFactoryFunction` to the result of calling the
```

File: index.html [hunk 8]
```diff
@@ -1819,65 +2021,53 @@ <h4>createDisclosureData</h4>
...
+and `selectivePointers`.
             </li>
             <li>
-Initialize `combinedFrame` to the result of calling the
-<a href="#jsonpointerstoframe"></a> primitive, passing `document` and the
-concatenation of `mandatoryPointers` and `selectivePointers` as `pointers`.
+Initialize `groupDefinitions` to a map with the following entries: key of
+the string "mandatory" and value of `mandatoryPointers`, key of the string
+"selective" and value of `selectivePointers`, and key of the string "combined"
+and value of `combinedPointers`.
```
Comment (TallTed) [inline index.html:2035]:
```suggestion
Initialize `groupDefinitions` to a map with the following entries: key of
the string `"mandatory"` and value of `mandatoryPointers`, key of the string
`"selective"` and value of `selectivePointers`, and key of the string `"combined"`
and value of `combinedPointers`.
```

File: index.html [hunk 9]
```diff
@@ -1819,65 +2021,53 @@ <h4>createDisclosureData</h4>
...
             <li>
-If `mandatoryFrame` and `combinedFrame` are both `null`, then throw an error
-indicating that nothing is to be disclosed.
+Initialize `groups` and `labelMap` to their associated values in the result
+of calling the algorithm in Section
+<a href="#canonicalizeandgroup"></a>, passing `document`,
+`labelMapFactoryFunction`, `groupDefinitions`, and any custom
+JSON-LD API options as parameters. Note: This step transforms the document into
+an array of canonical N-Quad strings with pseudorandom blank node identifiers
+based on `hmac` and groups the N-Quad strings according to selections based on
```
Comment (TallTed) [inline index.html:2044]:
```suggestion
based on `hmac`, and groups the N-Quad strings according to selections based on
```

File: index.html [hunk 10]
```diff
@@ -1903,9 +2093,35 @@ <h4>createDisclosureData</h4>
...
+Run the RDF Dataset Canonicalization Algorithm [[RDF-CANON]] on
+the joined <var>combinedGroup.deskolemizedNQuads</var>, passing any custom
+options, and get the canonical bnode identifier map, <var>canonicalIdMap</var>.
+Note: This map includes the canonical blank node identifiers that a verifier
+will produce when they canonicalize the reveal document.
+            </li>
+            <li>
+Initialize <var>verifierLabelMap</var> to an empty map. This map will map
+the canonical blank node identifiers the verifier will produce when they
+canonicalize the reveal document to the blank node identifiers that were
```
Comment (TallTed) [inline index.html:2109]:
```suggestion
canonicalize the revealed document to the blank node identifiers that were
```

File: index.html [hunk 11]
```diff
@@ -2191,30 +2413,27 @@ <h4>Base Proof Transformation (ecdsa-sd-2023)</h4>
...
             </li>
             <li>
-Initialize `nquads` to the result of calling the algorithm in Section
-<a href="#labelreplacementcanonize"></a>, passing `unsecuredDocument`, the
-result of calling the algorithm in Section
-<a href="#labelmapcanonize"></a> (passing `hmac`) as the
-`labelReplacementFunction`, and any custom JSON-LD API options. Note: This step
-transforms the document into an array of canonical N-Quads with pseudorandom
-blank node identifiers based on `hmac`.
+Initialize `labelMapFactoryFunction` to the result of callign the
```
Comment (TallTed) [inline index.html:2416]:
```suggestion
Initialize `labelMapFactoryFunction` to the result of calling the
```

File: index.html [hunk 12]
```diff
@@ -2191,30 +2413,27 @@ <h4>Base Proof Transformation (ecdsa-sd-2023)</h4>
...
             <li>
-Initialize `matching` and `nonMatching` to the result of calling the algorithm
-in Section <a href="#groupnquads"></a>, passing `nquads`, `mandatoryFrame` as
-`frame`, and any custom JSON-LD API options. Note: This step separates the
-N-Quads to mandatory (to disclose) and non-mandatory groups.
+Initialize `groups` to the result of calling the algorithm in Section
+<a href="#canonicalizeandgroup"></a>, passing `unsecuredDocument` as `document`,
+`labelMapFactoryFunction`, `groupDefinitions`, and any custom JSON-LD API
+options. Note: This step transforms the document into an array of canonical
+N-Quads with pseudorandom blank node identifiers based on `hmac` and groups
```
Comment (TallTed) [inline index.html:2428]:
```suggestion
N-Quads with pseudorandom blank node identifiers based on `hmac`, and groups
```

**Review by TallTed — CHANGES_REQUESTED** [review]

**Comment by dlongley** [comment]
@TallTed,

Thanks heaps! So I applied all of your suggestions except for the ones that changed the use of "and" when talking about passing parameters. Those suggest using "with" instead and I'm not sure that "with" is better?

Without that suggestion we have examples like this:

> Use algorithm X, passing foo as A and B.
> Use algorithm X, passing foo as A, bar as B, and C.

With that suggestion it would read like this:

> Use algorithm X, passing foo as A with B.
> Use algorithm X, passing foo as A, bar as B, with C.

I'm not sure what the consensus is, but "and" reads better to me. I can how the use of "with" reads alright when C is for example "any custom options":

> Use algorithm X, passing foo as A, bar as B, and any custom options.
vs.
> Use algorithm X, passing foo as A, bar as B, with any custom options.

But "and" still reads better to me personally and is consistent. Thoughts?

**Comment by TallTed** [comment]
Maybe...

Use algorithm X, passing A, and foo as B.
Use algorithm X, passing A, foo as B, and bar as C.
Use algorithm X, passing A, foo as B, and bar as C, along with any custom options.

**Comment by dlongley** [comment]
@TallTed,

> Maybe...
> 
> Use algorithm X, passing A, and foo as B.
> Use algorithm X, passing A, foo as B, and bar as C.
> Use algorithm X, passing A, foo as B, and bar as C, along with any custom options.

I'd be fine with that, but I'm also fine to just leave it as always using "and". If you want to adjust your suggestions accordingly, I'll add them in.

File: index.html [hunk 13]
```diff
@@ -1325,332 +1484,374 @@ <h4>jsonPointerToPaths</h4>
...
 
           <ol class="algorithm">
             <li>
-Initialize `skolemizedDocument` to the result of calling the algorithm in
-Section <a href="#toskolemizedjsonld"></a>, passing `nquads` and any custom
-JSON-LD API options (such as a document loader).
+Initialize <var>selectionDocument</var> to the result of the algorithm
+in Section <a href="#selectjsonld"></a>, passing
+<var>skolemizedCompactDocument</var> as <em>document</em>, with
+<var>pointers</var>.
```
Comment (TallTed) [inline index.html:1725]:
```suggestion
in Section <a href="#selectjsonld"></a>, passing <var>pointers</var>,
and <var>skolemizedCompactDocument</var> as <em>document</em>.
```

File: index.html [hunk 14]
```diff
@@ -1325,332 +1484,374 @@ <h4>jsonPointerToPaths</h4>
...
             </li>
             <li>
-Initialize `filteredDocument` to the result of calling the algorithm in Section
-<a href="#strictframe"></a>, passing `skolemizedDocument`, `filterFrame`, and
-any custom JSON-LD API options.
+Initialize <var>deskolemizedNQuads</var> to the result of the algorithm
+in Section <a href="#todeskolemizednquads"></a>, passing
+<var>selectionDocument</var> as <var>skolemizedCompactDocument</var>, with
```
Comment (TallTed) [inline index.html:1730]:
```suggestion
<var>selectionDocument</var> as <var>skolemizedCompactDocument</var>, and
```

File: index.html [hunk 15]
```diff
@@ -1325,332 +1484,374 @@ <h4>jsonPointerToPaths</h4>
...
+algorithm in Section <a href="#todeskolemizednquads"></a>, passing
+<var>skolemizedExpandedDocument</var> and any custom options.
             </li>
             <li>
-Initialize `matching` to a new map.
+Initialize <var>nquads</var> and <var>labelMap</var> to their associated
+values in the result of the algorithm in Section
+<a href="#labelreplacementcanonicalizenquads"></a>, passing
+<var>deskolemizedNQuads</var> as <var>nquads</var>,
+<var>labelMapFactoryFunction</var>, and any custom options.
```
Comment (TallTed) [inline index.html:1792]:
```suggestion
<var>labelMapFactoryFunction</var>, <var>deskolemizedNQuads</var> as
<var>nquads</var>, and any custom options.
```

File: index.html [hunk 16]
```diff
@@ -1325,332 +1484,374 @@ <h4>jsonPointerToPaths</h4>
...
             <li>
-Initialize filteredMatches to the values in `groupResult.matching`.
+For each key (<var>name</var>) and value (<var>pointers</var>) entry in
+<var>groupDefinitions</var>:
+              <ol class="algorithm">
+                <li>
+Add an entry with a key of <var>name</var> and a value that is the result
+of the algorithm in Section <a href="#selectcanonicalnquads"></a>, passing
+<var>skolemizedCompactDocument</var> as <var>document</var>,
+<var>pointers</var>, <var>labelMap</var>, and any custom options.
```
Comment (TallTed) [inline index.html:1805]:
```suggestion
<var>pointers</var>, <var>labelMap</var>, <var>skolemizedCompactDocument</var>
as <var>document</var>, and any custom options.
```

File: index.html [hunk 17]
```diff
@@ -1903,9 +2092,35 @@ <h4>createDisclosureData</h4>
               </ol>
             </li>
             <li>
+Initialize <var>revealDocument</var> to the result of the "selectJsonLd"
+algorithm, passing `document`, with `combinedPointers` as `pointers`.
```
Comment (TallTed) [inline index.html:2096]:
```suggestion
algorithm, passing `document`, and `combinedPointers` as `pointers`.
```

File: index.html [hunk 18]
```diff
@@ -2191,30 +2412,27 @@ <h4>Base Proof Transformation (ecdsa-sd-2023)</h4>
...
-Initialize `matching` and `nonMatching` to the result of calling the algorithm
-in Section <a href="#groupnquads"></a>, passing `nquads`, `mandatoryFrame` as
-`frame`, and any custom JSON-LD API options. Note: This step separates the
-N-Quads to mandatory (to disclose) and non-mandatory groups.
+Initialize `groups` to the result of calling the algorithm in Section
+<a href="#canonicalizeandgroup"></a>, passing `unsecuredDocument` as `document`,
+`labelMapFactoryFunction`, `groupDefinitions`, and any custom JSON-LD API
+options. Note: This step transforms the document into an array of canonical
+N-Quads with pseudorandom blank node identifiers based on `hmac`, and groups
+the N-Quad strings according to selections based on JSON pointers.
```
Comment (TallTed) [inline index.html:2428]:
```suggestion
<a href="#canonicalizeandgroup"></a>, passing `labelMapFactoryFunction`,
`groupDefinitions`, `unsecuredDocument` as `document`, and any custom JSON-LD
API options. Note: This step transforms the document into an array of canonical
N-Quads with pseudorandom blank node identifiers based on `hmac`, and groups
the N-Quad strings according to selections based on JSON pointers.
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Revised the `passing ... and` vs `passing ... with`

**Comment by Wind4Greg** [comment]
Hi Dave working on high level test vectors. Noticed redundant step to canonize proofOptions in section 3.5.3 Base Proof Hashing (ecdsa-sd-2023), this is already done in at the end of section 3.5.4 Base Proof Configuration (ecdsa-sd-2023)  -- step 9.

**Comment by dlongley** [comment]
@Wind4Greg,

Since there's considerable reworking done in this PR and still some editorial clean up to be done later, I'd like us to get this merged and then work off it as a baseline. That should include any simplifications like you mentioned.

**Review by msporny — APPROVED** [review]

**Review by TallTed — APPROVED** [review]

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**